### DISCUSSION TITLE
Slow performance using Zitadel authenticating services implemented in SpringBoot

### DISCUSSION THREAD

**Discussion by grewerto** [zitadel/zitadel#6220] [General]
Hi, we use Zitadel as cloud service to manage authentication/authorisation of backend services implemented in Springboot and a frontend client (react). For securing the backend services in Springboot we used an implementation suggested in Zitadel docs like:

`env.security.authorization-enabled=true`
`spring.security.oauth2.resourceserver.jwt.issuer-uri=https://xxx.zitadel.cloud/oauth/v2/token`
`spring.security.oauth2.resourceserver.jwt.jwk-set-uri=https://xxx.zitadel.cloud/oauth/v2/keys`
`spring.security.oauth2.resourceserver.opaquetoken.client-id=xxx`
`spring.security.oauth2.resourceserver.opaquetoken.client-secret=xxx`
`spring.security.oauth2.resourceserver.opaquetoken.introspection-uri=https://xxx.zitadel.cloud/oauth/v2/introspect`

The implementation works but the performance is quiet slow. A POST-request on introspect endpoint takes between 1700 to 2200 ms. I tried it out with a local Zitadel docker instance but the response times took still about  1000 ms. Do you have any ideas or suggestions how we could improve the performance?
Thank you

**Comment by fforootd** [comment]
Hi There

Usually the high latency you see comes down to the time zitadel uses to hash the passwords.
You have 3 options to improve this situation

1) Give zitadel more CPU 😁 
2) Decrease the hash cost https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L366 (11 could be an option and you need to create a new secret for this to have an effect)
3) Switch to private key authentication  https://zitadel.com/docs/apis/openidoauth/authn-methods#client-secret-basic

![image](https://github.com/zitadel/zitadel/assets/9879976/5eb5f129-875a-4c09-badb-dd92269fa468)

**Reply by ydemartino** [reply]
For anyone having the same problem, I used the third solution and the time went from 2.8s+ to 0.2s.
You can implement it using a custom `RequestEntityConverter` on the `SpringOpaqueTokenIntrospector`.

Here are some sample codes from my implementation (oauth2Properties is `OAuth2ResourceServerProperties`):
```
        fun createJwtToken(): Jwt {
            val claims = JwtClaimsSet.builder()
                .issuer(oauth2Properties.opaquetoken.clientId)
                .subject(oauth2Properties.opaquetoken.clientId)
                .audience(listOf(oauth2Properties.jwt.issuerUri))
                .expiresAt(Instant.now().plusSeconds(3_600L))
                .issuedAt(Instant.now())
                .build()
            return jwtEncoder.encode(JwtEncoderParameters.from(jwsHeader, claims))
        }
```
Then you need to set the converter:
```
        introspector.setRequestEntityConverter { token ->
            val headers = HttpHeaders()
            headers.contentType = MediaType.APPLICATION_FORM_URLENCODED
            val body = LinkedMultiValueMap<String, String>().apply {
                add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
                add("client_assertion", createJwtToken().tokenValue)
                add("token", token)
            }
            RequestEntity(body, headers, HttpMethod.POST, uri)
        }
```

**Reply by fforootd** [reply]
Thanks for sharing 😁
